cpp/src/arrow/acero/asof_join_node.cc

```diff
@@ -821,8 +817,11 @@ class InputState {
         ++batches_processed_;
         latest_ref_row_ = 0;
         have_active_batch &= !queue_.TryPop();
-        if (have_active_batch)
+        if (have_active_batch) {
           DCHECK_GT(queue_.UnsyncFront()->num_rows(), 0);  // empty batches disallowed
+          key_hasher_->Invalidate();  // batch changed - invalidate key hasher's cache
```

Comment:
Seems unrelated change - can you make a separate PR for this?

Reply:
Created #36094 and will revert here.